src/plots/cartesian/new_shape.js

```diff
@@ -0,0 +1,917 @@
+/**
+* Copyright 2012-2020, Plotly, Inc.
+* All rights reserved.
+*
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+
+
+'use strict';
+
+var parseSvgPath = require('parse-svg-path');
+
+var dragElement = require('../../components/dragelement');
+var dragHelpers = require('../../components/dragelement/helpers');
+var drawMode = dragHelpers.drawMode;
+var openMode = dragHelpers.openMode;
+
+var Registry = require('../../registry');
+var Lib = require('../../lib');
+var setCursor = require('../../lib/setcursor');
+
+var constants = require('./constants');
+var MINSELECT = constants.MINSELECT;
+var CIRCLE_SIDES = 32; // should be divisible by 4
+var i000 = 0;
+var i090 = CIRCLE_SIDES / 4;
+var i180 = CIRCLE_SIDES / 2;
+var i270 = CIRCLE_SIDES / 4 * 3;
+var cos45 = Math.cos(Math.PI / 4);
+var sin45 = Math.sin(Math.PI / 4);
+var SQRT2 = Math.sqrt(2);
+
+var helpers = require('./helpers');
+var p2r = helpers.p2r;
+var r2p = helpers.r2p;
+
+var handleOutline = require('./handle_outline');
+var clearOutlineControllers = handleOutline.clearOutlineControllers;
+var clearSelect = handleOutline.clearSelect;
+
+function recordPositions(polygonsOut, polygonsIn) {
+    for(var i = 0; i < polygonsIn.length; i++) {
+        var cell = polygonsIn[i];
+        polygonsOut[i] = [];
+        for(var j = 0; j < cell.length; j++) {
+            polygonsOut[i][j] = [];
+            for(var k = 0; k < cell[j].length; k++) {
+                polygonsOut[i][j][k] = cell[j][k];
+            }
+        }
+    }
+    return polygonsOut;
+}
+
+function displayOutlines(polygons, outlines, dragOptions, nCalls) {
+    if(!nCalls) nCalls = 0;
+
+    var gd = dragOptions.gd;
+
+    function redraw() {
+        // recursive call
+        displayOutlines(polygons, outlines, dragOptions, nCalls++);
```

Comment:
what's `nCalls`? Doesn't seem used. Debugging?

Reply:
Good Call. Now used in https://github.com/plotly/plotly.js/pull/4775/commits/c5ff5dfeec859a1c46cfc1494155bab8a1270bd0.